scripts/checkpoint_averaging/torch_distributed_checkpoint_averaging.py

```diff
@@ -0,0 +1,253 @@
+import tarfile
+
+import numpy as np
+import tensorstore  # need to import it for bf16 support
+import torch
+import torch.distributed as dist
+import zarr
+
+try:
+    from megatron.core.dist_checkpointing import load, save
+    from megatron.core.dist_checkpointing.mapping import ShardedObject, ShardedTensor
+    from megatron.core.dist_checkpointing.serialization import load_sharded_metadata
+    from megatron.core.dist_checkpointing.strategies.torch import TorchDistSaveShardedStrategy
+except ImportError:
+    raise ImportError("Megatron is not installed. Please install Megatron-LM to use this script.")
+
+from tqdm import tqdm
+
+
+"""
+Script to average distributed checkpoints and save new .nemo checkpoint.
+
+Example: python scripts/checkpoint_averaging/torch_distributed_checkpoint_averaging.py \
+                --name_prefix=<checkpoint name> \
+                --checkpoint_dir=<folder with .discp files> \
+                --untarred_nemo_folder=<path to the untarred nemo checkpoint to get config and tokenizers> \
+                --steps <optinally a list of checkpoint steps to average, if not provided, it will average all the checkpoints>
+"""
+
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s - %(levelname)s - %(message)s",
+    datefmt="%Y-%m-%d %H:%M:%S",
+)
+
+
+def init_distributed():
+    if not dist.is_available():
+        raise ValueError("Distributed package is not available in your PyTorch build.")
+
+    if not dist.is_initialized():
+        logging.info(
+            "Initializing distributed process group: rank %s, world_size %s",
+            os.environ["LOCAL_RANK"],
+            os.environ["WORLD_SIZE"],
+        )
+        dist.init_process_group(backend="gloo")
+    else:
+        logging.info("Distributed process group already initialized.")
+
+
+def cleanup_distributed():
+    """
+    Cleans up the distributed process group.
+    """
+    dist.destroy_process_group()
+
+
+def load_distributed_weights_dcp(ckpt_path):
+    logging.info(f"Starting to load checkpoint from {ckpt_path}")
+    metadata = load_sharded_metadata(ckpt_path)
+
+    keys_to_delete = [k for k in metadata if k.startswith("optimizer.")]
+    for k in keys_to_delete:
+        del metadata[k]
+
+    state_dict = load(
+        sharded_state_dict=metadata,
+        checkpoint_dir=ckpt_path,
+        validate_access_integrity=True,
+    )
+    model_state_dict = {}
+    for k, v in state_dict.items():
+        if k.startswith("model."):
+            model_state_dict[k] = v
+    logging.info(f"Loaded checkpoint from {ckpt_path}")
+    return model_state_dict, metadata
+
+
+def save_distributed_weights_dcp(ckpt_path, averaged_state_dict):
+    logging.info(f"Saving averaged checkpoint to {ckpt_path}")
+
+    os.makedirs(ckpt_path, exist_ok=True)
+    save(
+        sharded_state_dict=averaged_state_dict,
+        checkpoint_dir=ckpt_path,
+        sharded_strategy=TorchDistSaveShardedStrategy(backend="torch_dist", version=1),
```

Comment:
You can just pass `sharded_strategy=torch_dist` and consequently make it configurable in the CLI (allowing e.g. 'zarr' if someone wants to have 'zarr' for some reason; it's deprecated but sometimes people fallback to it)

Reply:
There's a script for zarr format [here](https://github.com/NVIDIA/NeMo/blob/main/scripts/checkpoint_averaging/distributed_checkpoint_averaging.py).
I guess it can be merged into one, but so far, all of the formats had separate scripts

Reply:
I think having one script instead of 2 will be better going forward, but I'm not a user of those scripts so maybe there are some details I'm missing.

We don't have to remove the other one, we could just add an option to this one for a smooth transition